### DISCUSSION TITLE
Smart build system detection

### DISCUSSION THREAD

**Discussion by richard-hajek** [go-task/task#1116] [Ideas]
A fun quality of life addition would be for `task` to function as an alias to existing build systems. For example, when encountering `package.json` `task` would delegate all calls to `npm`, unless an existing Taskfile would be present.

Possible use cases:
- Translating calls from `task` to `npm`
- Translating calls from `task` to `cargo`
- Translating calls from `task` to `go`
....

This would be by no means a full replacement of the original tool. The `task build` would do little more than an alias to `go build`/`cargo build`/`npm build`. Implementation wise it could be basically a list of known build systems and a Taskfile for each of them.

**Comment by pd93** [comment]
This might cause issues when a project contains both a `go.mod` and ` package.json`. For example, if I run `task test` do we call `go test` or `npm test`? How do we decide which one gets priority? I think most users would just end up creating wrapper tasks in a `Taskfile` if they wanted this functionality to be exposed via the `task` command. This ensures they're always going to be running the correct command and gives them much better control.

Using a wrapper also lets you add more advanced functionality. For example, instead of `task install` just exposing `npm install`, we can create a custom wrapper and add fingerprinting so that it only gets run if the `package.json` has been updated or if `node_modules` is missing.

**Comment by richard-hajek** [comment]
> This might cause issues when a project contains both a go.mod and  package.json. For example, if I run task test do we call go test or npm test? How do we decide which one gets priority?

Give up with a text message "No Taskfile found and ambigous project type, we are unable to find reasonable command"

> I think most users would just end up creating wrapper tasks in a Taskfile if they wanted this functionality to be exposed via the task command.

True. However the usecase I had in mind would be

1. You download a random NPM project
2. You don't feel like checking NPM docs for the specific build command (is it `npm build`, `node build`, `npm run build`? )
3. You run task build, which will print the actual command it uses, so you know what to use next time instead, and it also does the commnd

> Using a wrapper also lets you add more advanced functionality. 

All of this could also be included in the default built-in wrapper functionality. But that would require expert knowledge in all those other build systems which might be overkill and too much work

**Comment by andreynering** [comment]
Hi @richard-hajek,

I think your idea is valid but it does not fit this project IMO. Task should be focused on running Taskfiles.

I think your idea could implemented on a brand new project, dedicated to this.
